Added push hook that calls the linter and front/backend test.

brunnre8 opened the pull request:
Modified CONTRIBUTING.md to notify new users about this.
Added installation to install_third_party.sh

seanlip commented:
/cc @MAKOSCAFEE, @kevinlee12

wxyxinyu commented:
Hi @brunnre8, I took a quick look at how this is working, and noticed some things:
- the first time I try to push after installing the script, the linter will lint all files. On subsequent runs, it behaves as expected
- I seem to be unable to do a `git push origin --delete [branch]`. I get this error:
```ref_list:
[GitRef(local_ref='(delete)', local_sha1='0000000000000000000000000000000000000000', remote_ref='refs/heads/lint_test_push_hook', remote_sha1='232e22f69961dad2d0556d9e35228060eccf864e')]
fatal: ambiguous argument 'origin/(delete)': unknown revision or path not in the working tree.
```
- if I accidentally try to push without making a commit, I get an error message:
```ref_list:
[]
Files being pushed:
[]
usage: pre_commit_linter.py [-h] [--path PATH | --files FILES [FILES ...]]
pre_commit_linter.py: error: argument --files: expected at least one argument
Push failed, please correct the linting issues above
error: failed to push some refs to 'https://github.com/wxyxinyu/oppia.git'
```
instead of a `Everything up-to-date` that git gives by default

Of these, I think it's important that the second point is fixed before this is merged.

Thanks!

seanlip commented:
Hi @brunnre8, I have one other question: is there an override mechanism for this (that bypasses the checks)? If so, how is it invoked? I suspect there may be occasional cases in which such a mechanism might be needed, so it might be a good idea to document it at the top of the script.

seanlip commented on scripts/pre_push_hook.py:
```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python2.7
```
Could you add copyright etc. headers, same as for the other scripts in this folder?

brunnre8 replied:
I can and I have :smiley:

kevinlee12 commented on scripts/pre_push_hook.py:
```diff
@@ -0,0 +1,184 @@
+_PARSER.add_argument("url", nargs="?", help="provided by git before push")
+_PARSER.add_argument("--install", action="store_true", default=False,
+                     help="Install pre_push_hook to the .git/hooks directory")
+_ARGS = _PARSER.parse_args()
```
This is probably better placed in main because that's closer to where we are using it.

brunnre8 replied:
fixed

kevinlee12 commented on scripts/pre_push_hook.py:
```diff
@@ -0,0 +1,184 @@
+_PARSER.add_argument("--install", action="store_true", default=False,
+                     help="Install pre_push_hook to the .git/hooks directory")
+_ARGS = _PARSER.parse_args()
+REMOTE = _ARGS.remote
```
Same as comment above (place in main() )

brunnre8 replied:
also addressed

kevinlee12 commented on scripts/pre_push_hook.py:
```diff
@@ -0,0 +1,184 @@
+        ValueError if git command fails
+    """
+    remote_branch = remote_branch if remote_branch else local_branch
+    git_remote = "{}/{}".format(remote, remote_branch)
```
while `"{}/{}"` works, might be clearer if it were ` "{0}/{1}" `

seanlip replied:
Actually, could we use '%s/%s' % (..., ...)? We generally use string interpolation by convention, in this codebase.

brunnre8 replied:
Fixed that

kevinlee12 commented on scripts/pre_push_hook.py:
```diff
@@ -0,0 +1,184 @@
+    pprint.pprint(modified_files)
+    lint_status = _start_linter(modified_files)
+    if lint_status != 0:
+        sys.exit("Push failed, please correct the linting issues above")
```
Might be better if we do a print of the message, then `sys.exit`. @seanlip, second opinion please.

seanlip replied:
Yes, I agree. This is what we do elsewhere, so we should be consistent.

brunnre8 replied:
The next version has all the exit messages as separate print statements

seanlip commented on scripts/pre_push_hook.py:
```diff
@@ -0,0 +1,184 @@
+import shutil
+
+_PARSER = argparse.ArgumentParser()
+_PARSER.add_argument("remote", nargs="?", help="provided by git before push")
```
Bit of a nit, but would you mind using single quotes for strings throughout this file? We do this by convention throughout the codebase.

brunnre8 replied:
I do not no, I was just using the standard ones from my locale as a force of habit.
Fixed

brunnre8 commented:
@wxyxinyu 
>the first time I try to push after installing the script, the linter will lint all files. On subsequent runs, it behaves as expected

This will only happen if there is no `REMOTE/develop` branch in the remote being pushed to (even if it's local). The order of lookups is the following:
     1. Compare `BRANCH` to `REMOTE/BRANCH` - perfect solution
     2. Compare `BRANCH` to `REMOTE/develop` - fallback strategy as to not lint unnecessarily
     3. There is no existing branch to reference to -  give up consider all files "new"

>I seem to be unable to do a git push origin --delete [branch]

Indeed, I forgot about deleting branches, fixed

>If I accidentally try to push without making a commit, I get an error message

Also fixed. A question there, running the front-/backend tests is still a good thing to do yes? It will not try and lint stuff as there are no new files, but testing can't hurt in my opinion. Your thoughts?

@seanlip 
> Is there an override mechanism for this ?

Yes, again with `git push REMOTE BRANCH --no-verify`.
I put that into the `CONTRIBUTING.md` as well.
I'm kind of reluctant to make it too easy for people to circumvent that (e.g. by providing it in the "exit-log" message as this will most probably have the undesired effect of frustrated people just bypassing the hook (after having the linter complain a few times at you I'm sure you can relate).
Do you disagree on that? I can easily provide it into the error messages.

wxyxinyu commented:
For the first point, I can't seem to reproduce it; not sure what happened there. 

If someone attempts to push with no modified files, I think we should just abort with an error message, since it probably means they forgot to commit the changes. This is the default behavior too (getting an "Everything up-to-date" message).

seanlip commented:
@brunnre8: Re the override mechanism, I definitely agree that circumvention should not be too easy, and that the message shouldn't go in the exit log. In addition, I don't even think you need to put it into CONTRIBUTING.md -- just the docstring at the top of the file should be enough, I think.

Also, thanks for addressing the comments! Just for clarity, @wxyxinyu will be the one giving final merge approval on this PR.

brunnre8 commented:
@seanlip Yes sure, I just replied to you as you brought the issue up.

@wxyxinyu , @seanlip about the addition to CONTRIBUTING.md, I thought that may be handy if there is someone who needs to get non working code out to other contributors.

The issue is, that the hook is installed in the background without much notice. So people will probably not  know where to look for the solution. At least in my opinion putting an entry in one of the wiki-pages may be worth it. What would you like me to do? 

@wxyxinyu regarding not testing when it's not needed.
This may take me a while as I have to refactor the meat of my script in order to do that.
The issue being that one can do really fancy stuff with just one push command. For example:
* deleting a branch on the remote
* pushing another branch to the remote
* pushing a branch which doesn't have any hanging commits
* pushing a branch that only deletes some files on the repo (and I really do not want to lint deleted files)

And that all in one command (who would ever do that, but the point being that one could. So we have to address it)

wxyxinyu commented:
@brunnre8 I think it is fine to include it in CONTRIBUTING.md

Regarding not testing when it's not needed, I was thinking of the specific case of trying to push when there are no changes. Will it be possible to not run tests if there are no changes? Of course, doing complicated stuff on one push command should work too, but linting/tests should be run then.

brunnre8 commented:
@wxyxinyu All the issues mentioned above should be addressed.

There is a severe issue however that might make it much more interesting...
I just pushed the second version via the hook (meaning I actually let it run through). Github was kind enough to close the connnection about midway through the tests... it then exits (after the tests are run) without any error messages, but it also didn't push!

What this means is that basically our test suit runs too long and runs into Githubs server timeout.

What we could try to do is abort in the hook and then try and push from there (with the --no-verify flag) however this seems rather unstable to me. Thoughts?

Oh and as a word of caution: As the hook relies on other scripts we have to make sure that they exit with an non-zero exit code if the linting  or some tests fail, just printing stuff to the console will not be enough anymore.

/CC @seanlip @kevinlee12 
Any thoughts about the time out issue guys?

seanlip commented:
@brunnre8 -- hm, that's interesting. Really, what we should do is to execute the tests etc. _before_ opening the connection to Github. Do you know if there's a way to do that?

seanlip commented:
Here's an interesting workaround: http://stackoverflow.com/a/33130116

brunnre8 commented:
Yap, that's my idea as well... however this seems quite unstable as the user might already do some other stuff in another terminal window and if the person changed the branch while doing that and I initiate a `git push` that might be rather unlucky.

Of course one could do it a little bit more fine grained by reading out the information properly but how easy that is is the other question

Sadly opening up a connection after running the test suit is not feasible as the push hook is run *after* the remote has been initialized (in the end that's the point of it).

Would just linting be an option? The test suit is anyway run by Travis if I understood correctly yes?

seanlip commented:
Actually, both linting and the test suite are run by Travis. The pre-push hook is therefore more of a convenience: it would be nice to deal with the errors while the user is still working on the local machine, I think.

So, two ideas:
* Just linting could be an option, as you suggest. Maybe linting + frontend tests, since the latter are fast.
* We could create a script that the developer runs instead of "git push" -- something like "python scripts/safe_push.py". That will run the checks and then call "git push" from within it. Developers have the choice of which one to use.

What do you think? It might be a good idea to ask other developers, too, what their preferred workflow would be. Maybe make a post on oppia-dev@ or the gitter chat room? Also, I'm /cc ing @wxyxinyu, @kevinlee12, @MAKOSCAFEE and @BenHenning to request their opinions.

kevinlee12 commented:
Personally, I would prefer a script that runs as suggested by Sean, linting and frontend, that way, I have more control over what to do in case tests fail.

brunnre8 commented:
@kevinlee12 would you mind elaborating your answer?
I'm not quite sure if you meant his first idea (pre-push hook with linting/front-end tests) or his second one (secondary "git push like" script that runs the linting and *all* the tests)

kevinlee12 commented:
@brunnre8, sure, first idea (pre-push hook with linting/front-end tests)

brunnre8 commented:
@seanlip I would prefer your first idea (pre-push hook with linting/frontend) as implementing a different script to push would somehow screw with the workflow people are used to with git.
But I think asking around might be a good idea

seanlip commented:
OK, could you send an email to the oppia-dev@ mailing list? It's at: https://groups.google.com/forum/#!forum/oppia-dev

BenHenning commented:
I'm not sure I'm on-board with a pre-push hook, at all. Whether we do a pre-push hook for linting or circumvent the ``git push`` workflow and provide our own script seems to complicate the process of ultimately pushing code to the forks. Beyond that, I'm uncomfortable with a script changing the code I'm working on without my explicit knowledge. Sure, developers will quickly learn that this process is going to alter their code, but they aren't explicitly activating that feature; it's being forced.

I would really rather have Travis report linter errors (as it does now) which prevent a merge into develop, and then have a script which is manually run which goes over the codebase and attempts to fix the linter errors automatically. Linting really should never be automatic, I think. Perhaps this isn't the answer you're looking for, though. :)

brunnre8 commented:
@BenHenning we do not change the code. 
What we refer to as 'linting'  is just checking if the code adheres to the guidelines. It will not change anything 

We would basically do locally what travis does remotely.  The nice thing about that is,  that you get feedback while you're still worjing on it and not several minutes/hours later when one gets the email from github

BenHenning commented:
Ah, that's quite interesting. Thanks for the clarification, @brunnre8. I'm sort of interested in why we're introducing a new developer pattern, here. In the past, we run the test scripts manually if we wish to figure out whether we are passing tests. Why not run the linter script manually for the same purpose?

I do see benefit in making it part of the required developer pipeline, but I am really concerned this is going to overcomplicate the developer pipeline too much. I'm sort of torn on this, so I'm going to refrain from making a final opinion until I see others' arguments.

seanlip commented:
@BenHenning -- I'm not sure it's a new developer pattern, actually, in the sense that what the developer does still remains the same. It's just that (at least for the first idea I had), the "git push" process takes a bit longer. Can you elaborate more on your concerns with regards to overcomplication?

Personally, I actually do like the idea of an auto-presubmit check because, as a developer, I sometimes forget to run the lint script, and if I leave this to Travis-CI to figure out, the lag time is quite long and I'll have moved on to doing other things (as @brunnre8 says). So, having some of these checks happen automatically as part of the regular dev workflow seems quite nice to me. The only reason why we're not including the backend and e2e tests in this is because they take too long, but we would if they ran faster.

Also, I should ask: @brunnre8 -- as a developer, do you have a preference, too?

brunnre8 commented:
@seanlip well I'm paranoid... I tend to run test frequently before I commit if they are available :wink: 
But having the files checked locally upon push sounds reasonable to me,  I would definitely not do a pre-commit hook but pre-push makes sense.

wxyxinyu commented:
I think having a pre-push hook is good because it doesn't affect the steps in the current current development process, but it also has a check to ensure people don't forget to lint/run tests. 

@brunnre8 I think the first option, just run linting/frontend tests is good. That way people don't have to modify their current process, and won't just forget to run the new script.

seanlip commented:
I think I'm seeing most of the opinions in support of a pre-push hook with lint + frontend tests, so, @brunnre8, let's go with that for now. We can change it to something else later if it doesn't work in practice for any reason.

Btw, looks like your branch has merge conflicts with develop -- these need to be resolved before the PR can be merged. You can do so by merging develop into your branch.

brunnre8 commented:
@wxyxinyu In case it looks good to you, please give me a short heads-up so I can send a notification to all the developers what's going to happen once they pull (or alternatively send one yourself).

Having a hook installed in the background without people knowing that it's gonna happen is something that we should avoid in my opinion.

The script now runs just the linting plus the frontend-tests as discussed.

seanlip commented:
Ah -- I don't use Windows, I'm afraid. I think @varenya does though, and if I'm not mistaken @BenHenning and @kevinlee12 have access to Windows machines (though I might be wrong in @kevinlee12's case).

seanlip commented:
(Also, FYI @brunnre8, if you edit a comment, GitHub doesn't send an email notification -- I noticed the last paragraph only by chance.)

wxyxinyu commented:
@brunnre8 I seem to still be unable to delete branches. Does it work on your side?

kevinlee12 commented:
sorry, no access to a windows machine

brunnre8 commented:
@wxyxinyu It was before I messed up the indention level...
I'm terribly sorry for wasting your time, I really thought I checked it all again before pushing...

Now it should finally be working (I triple checked this time) :persevere:

wxyxinyu commented:
@brunnre8 Not a problem! Everything looks good to me now. 

I think this will be ok on Windows because we run the system on a VM there, but do you think it's worth checking?

seanlip commented:
FYI, I expect the Windows process to be fine; our current Windows installation process essentially sets up a Linux VM. So I don't think it needs additional testing.

brunnre8 commented:
@seanlip hm I beg to differ.... if varenya is right and vagrant doesn't allow symlinks then it wont be the same as in linux (this being what I am essentially doing) 
for windows what it does right now is just copy the hook every time the setup script is called (and I'm not quite happy with that) 
Another option would be to hash the file to check if it changed,  however I have no experience with that. 
Your thoughts? 

@wxyxinyu I still have my old laptop with Windows laying around... I'll try and test it out this evening

seanlip commented:
Hm, fair point. I'm not sure what's causing the symlinks issue though; @philoniare did a windows installation successfully and didn't run into that problem. But if you're able to test it out in a Windows machine, that would be great too.

I just wanted to strike a balance between that holding up this push too much, since I think I've now seen two or three cases where contributors forgot to run the linter before making the push -- so, hopefully we can get this in within the next couple of days!

wxyxinyu commented:
@brunnre8 based on discussion on gitter, could you please add a comment saying that this pre-push hook will do nothing on windows? After that this can be merged. Thanks!

wxyxinyu commented:
@brunnre8 -- This is great, thanks! Merging. 

By the way, I don't get notifications when you push new commits, and only saw this by chance. Could you reply to indicate that changes have been pushed next time? Thanks!

BenHenning commented:
I do have access to a Windows machine, but it's running under Git for Windows which basically runs in a MinGW Unix-like environment. I don't actively develop on Windows so it's unlikely that I test this in the near future. :) But, as @seanlip mentioned, it probably won't plague any of our actual Widows users due to setting up a VM instead.
